### DISCUSSION TITLE
Hiding lines from code blocks

### DISCUSSION THREAD

**Discussion by ollipa** [squidfunk/mkdocs-material#7113] [Q&A] [answered]
Is there a way to hide lines in the rendered code blocks? I'm running the code blocks as a part of my doctest test suite and it would be useful to have some setup/teardown code in the code blocks that could be hidden from the rendered documentation.

Right now I'm adding my teardown code in a comment after the code block:
```
\```python
>>> mocker(Teapot).mock("boil").called_once()
<chainmock._api.Assert object at ...>
>>> Teapot().boil()

\```

<!--
# Reset mocks between examples
>>> State.teardown()

-->
```

However, it would be nice if I could do something like this:
```
\```{.python hide_lines="4"}
>>> mocker(Teapot).mock("boil").called_once()
<chainmock._api.Assert object at ...>
>>> Teapot().boil()
>>> State.teardown()

\```
```

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello @ollipa,
after consulting the SuperFences and Highlight docs I don't see any option to hide lines:
- https://facelessuser.github.io/pymdown-extensions/extensions/superfences/
- https://facelessuser.github.io/pymdown-extensions/extensions/highlight/

However, after looking at the source code, I see that `hl_lines` is passed to the `Highlight.highlight` function together with the `src` code to be highlighted. So you could create a [`hook`](https://www.mkdocs.org/user-guide/configuration/#hooks) which adds a [decorator](https://www.freecodecamp.org/news/the-python-decorator-handbook/) function to monkey-patch either [`Highlight.highlight`](https://github.com/facelessuser/pymdown-extensions/blob/8ed2fcb8ac688452ae1f5b29ab403e1e1d97e4e5/pymdownx/highlight.py#L326) or [`‎SuperFencesBlockPreprocessor.highlight`](https://github.com/facelessuser/pymdown-extensions/blob/8ed2fcb8ac688452ae1f5b29ab403e1e1d97e4e5/pymdownx/superfences.py#L739) to handle your own `hide_lines` attribute and remove the lines yourself. The hook needs to patch the function before it's invoked, so likely the [`on_startup`](https://www.mkdocs.org/dev-guide/plugins/#on_startup) event is the best option, as it runs only once.

**Reply by ollipa** [reply]
Thank you! I think I'll try the monkey patch approach.

Do I understand correctly that without monkey patching this feature would need to be implemented in Pygments and Highlight extension?

**Reply by kamilkrzyskow** [reply]
Pygments is a lexer, it tokenizes the given code, so I don't think removing lines is in scope for them. 

Highlight or some other PyMdown Extension influencing the SuperFences would have to implement that. Also note that removing lines from the `src` does create an issue with `hl_lines`, as it would have to "know" about the removal of lines and dynamically shift the given indexes for highlighted lines, as otherwise it would select wrong lines. Also if removing the lines, will remove a open/close token in the code it might break Pygments highlighting.

Another option is to extract the code block contents into another file and use Snippets, as it has the option to select the lines of code per number or per named section: 
- https://facelessuser.github.io/pymdown-extensions/extensions/snippets/

Consider sharing your solution for others wanting to have the same feature.

**Reply by ollipa** [reply]
I was able to implement the hook. Here is my implementation in case it is useful to someone:
https://github.com/ollipa/chainmock/commit/ebc425da3193a1740376693fafed4559da2cc75b

Using Snippets is a good option too but I prefer to keep the code examples in the markdown file. Thank you again, you were very helpful.
